apollo-router/src/cache/storage.rs

```diff
@@ -82,9 +82,10 @@ where
     }
 
     pub(crate) async fn get(&self, key: &K) -> Option<V> {
-        let mut guard = self.inner.lock().await;
         let instant_memory = Instant::now();
-        match guard.get(key) {
+        let res = self.inner.lock().await.get(key).cloned();
+
+        match res {
```

Comment:
Won't this introduce the kind of racing we are trying hard to avoid?

e.g.: If the key isn't present, then a client will get `None` and immediately release the lock. Lots of client could all call in parallel and get the same result. Then they will all (as a "thundering herd") call to redis to get the key...

Reply:
calls to the APQ and query planner cache, which use the dual in memory and redis cache, are gated by the deduplication code

Reply:
If they are gated, doesn't that mean we don't need the change?

Reply:
the lock is on the entire in memory cache, not per key. There can be multiple requests with different keys, so deduplication will let through one request per different key, and so if the first key got a cache miss and then queries redis, the request for the second key should not be blocked on the mutex of the in memory cache